src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java

```diff
@@ -0,0 +1,229 @@
+package com.google.devtools.build.lib.vfs;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.hash.Funnel;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.security.DigestException;
+import java.security.MessageDigest;
+
+public final class Blake3MessageDigest extends MessageDigest implements Hasher {
```

Comment:
If you use `MessageDigestHashFunction` as suggested below, you can remove `implements Hasher`.

Reply:
Sorry, I missed that. Can we then create a separate class to `implement Hasher`? I found it is confusing to combine these two especially when we have `update`, `engineUpdate` and share the underlying buffer.

Reply:
Done.